cupyx/scipy/interpolate/_interpolate.py

```diff
@@ -0,0 +1,972 @@
+
+import cupy
+from cupy._core import internal  # NOQA
+from cupy._core._scalar import get_typename  # NOQA
+from cupyx.scipy import special as spec
+from cupyx.scipy.interpolate._bspline import BSpline
+
+import numpy as np
+
+
+TYPES = ['double', 'thrust::complex<double>']
+INT_TYPES = ['int', 'long long']
+
+INTERVAL_KERNEL = r'''
+#include <cupy/complex.cuh>
+
+#define le_or_ge(x, y, r) ((r) ? ((x) < (y)) : ((x) > (y)))
+#define ge_or_le(x, y, r) ((r) ? ((x) > (y)) : ((x) < (y)))
+#define geq_or_leq(x, y, r) ((r) ? ((x) >= (y)) : ((x) <= (y)))
+
+extern "C" {
+__global__ void find_breakpoint_position(
+        const double* breakpoints, const double* x, long long* out,
+        bool extrapolate, int total_x, int total_breakpoints, bool asc) {
+
+    int idx = blockDim.x * blockIdx.x + threadIdx.x;
+    if(idx >= total_x) {
+        return;
+    }
+
+    double xp = *&x[idx];
+    double a = *&breakpoints[0];
+    double b = *&breakpoints[total_breakpoints - 1];
+
+    if(isnan(xp)) {
+        out[idx] = -1;
+        return;
+    }
+
+    if(le_or_ge(xp, a, asc) || ge_or_le(xp, b, asc)) {
+        if(!extrapolate) {
+            out[idx] = -1;
+        } else if(le_or_ge(xp, a, asc)) {
+            out[idx] = 0;
+        } else if(ge_or_le(xp, b, asc)) {
+            out[idx] = total_breakpoints - 2;
+        }
+        return;
+    } else if (xp == b) {
+        out[idx] = total_breakpoints - 2;
+        return;
+    }
+
+    int left = 0;
+    int right = total_breakpoints - 2;
+    int mid;
+
+    if(le_or_ge(xp, *&breakpoints[left + 1], asc)) {
+        right = left;
+    }
+
+    bool found = false;
+
+    while(left < right && !found) {
+        mid = ((right + left) / 2);
+        if(le_or_ge(xp, *&breakpoints[mid], asc)) {
+            right = mid;
+        } else if (geq_or_leq(xp, *&breakpoints[mid + 1], asc)) {
+            left = mid + 1;
+        } else {
+            found = true;
+            left = mid;
+        }
+    }
+
+    out[idx] = left;
+}
+}
+'''
+
+INTERVAL_MODULE = cupy.RawModule(
+    code=INTERVAL_KERNEL, options=('-std=c++11',),)
+
+PPOLY_KERNEL = r"""
+#include <cuda_runtime.h>
```

Comment:
Would you use `hip_runtime.h` instead of `cuda_runtime.h` on ROCm? You can use `cupy_backends.cuda.api.runtime.is_hip` to find if the environment is CUDA or ROCm.

Reply:
ah, I'll make the fix!